### PR TITLE
[CH-182] Support soft affinity scheduling for Gluten

### PR DESCRIPTION
**Pull request by zzcclp**
Support soft affinity scheduling for Gluten.

## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

File: shims/common/src/main/scala/io/glutenproject/GlutenConfig.scala [hunk 1]
```diff
@@ -262,6 +262,15 @@ object GlutenConfig {
   val GLUTEN_CLICKHOUSE_SEP_SCAN_RDD_DEFAULT = true
   val GLUTEN_CLICKHOUSE_CONFIG_PREFIX = "spark.gluten.sql.columnar.backend.ch"
 
+  // For Soft Affinity Scheduling
+  val GLUTEN_SOFT_AFFINITY_ENABLED = "spark.gluten.soft-affinity.enabled"
+  val GLUTEN_SOFT_AFFINITY_ENABLED_DEFAULT_VALUE = false
+  val GLUTEN_SOFT_AFFINITY_REPLICATIONS_NUM = "spark.gluten.soft-affinity.replications.num"
```
Comment (zhouyuan) [inline shims/common/src/main/scala/io/glutenproject/GlutenConfig.scala:268]: Can you please help to write a short note on how to use this? does this mean by providing two "fake" replicas, then spark will schedule to the first place?

Comment (zzcclp) [reply shims/common/src/main/scala/io/glutenproject/GlutenConfig.scala:270]: add the doc for these configs, please take a review

**Comment by FelixYBW** [comment]
@zhouyuan  is it related to the HDFS remote read issue we noted?

**Comment by zhouyuan** [comment]
> @zhouyuan is it related to the HDFS remote read issue we noted?

per my understanding, this patch is trying to solve the affinity issue on compute disaggregate storage case

**Comment by zzcclp** [comment]
> > @zhouyuan is it related to the HDFS remote read issue we noted?
> 
> per my understanding, this patch is trying to solve the affinity issue on compute disaggregate storage case

Currently, the preferred locations of Gluten's NativePartition is empty, maybe this pr can fix this issue.

File: jvm/src/main/scala/org/apache/spark/softaffinity/SoftAffinityUtil.scala [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+          s"the expected executors are ${locations.mkString("_")} ")
+        locations.map { p =>
+          if (p._1.equals("")) p._2
+          else ExecutorCacheTaskLocation(p._2, p._1).toString
+        }.toArray
+      } else {
+        Array.empty[String]
+      }
+    } else {
+      expectedTargets
```
Comment (zhouyuan) [inline jvm/src/main/scala/org/apache/spark/softaffinity/SoftAffinityUtil.scala:55]: note: this will help on Spark + HDFS co-located case

**Review by zhouyuan — APPROVED** [review]

**Pull request merged**